runtime/vm/criuhelpers.cpp

```diff
@@ -103,4 +105,259 @@ isCheckpointAllowed(J9VMThread *currentThread)
+	pool_state walkState = {0};
+	BOOLEAN result = TRUE;
+	J9VMThread *currentThread = vm->mainThread;
+
+	J9InternalHookRecord *hookRecord = (J9InternalHookRecord*)pool_startDo(hookRecords, &walkState);
+	while (NULL != hookRecord) {
+		J9Class *clazz = J9OBJECT_CLAZZ_VM(vm, object);
+		if ((clazz == hookRecord->instanceType)
+			|| (hookRecord->includeSubClass && isSameOrSuperClassOf(hookRecord->instanceType, clazz))
+		) {
+			if (NULL == hookRecord->instanceObjects) {
+				hookRecord->instanceObjects = pool_new(sizeof(J9Object), 0, 0, 0, J9_GET_CALLSITE(), OMRMEM_CATEGORY_VM, POOL_FOR_PORT(vm->portLibrary));
+				if (NULL == hookRecord->instanceObjects) {
+					setNativeOutOfMemoryError(currentThread, 0, 0);
+					result = FALSE;
+					break;
+				}
+			}
+			J9Object *objectRecord = (J9Object*)pool_newElement(hookRecord->instanceObjects);
+			if (NULL == objectRecord) {
+				result = FALSE;
+				setNativeOutOfMemoryError(currentThread, 0, 0);
+			} else {
+				objectRecord->clazz = (j9objectclass_t)object;
+			}
+		}
+
+		hookRecord = (J9InternalHookRecord*)pool_nextDo(&walkState);
+	}
+
+	return result;
+}
+
+static jvmtiIterationControl
+objectIteratorCallback(J9JavaVM *vm, J9MM_IterateObjectDescriptor *objectDesc, void *userData)
+{
+	j9object_t object = objectDesc->object;
+
+	fillinHookRecords(vm, object);
+
+	return JVMTI_ITERATION_CONTINUE;
+}
+
+BOOLEAN
+runInternalJVMCheckpointHooks(J9JavaVM *vm)
+{
+	J9VMThread *currentThread = vm->mainThread;
+	J9Pool *hookRecords = vm->checkpointState.hookRecords;
+	pool_state walkState = {0};
+	BOOLEAN result = TRUE;
+
+	Trc_VM_criu_runCheckpointHooks_Entry(currentThread);
+
+	/* Iterate heap objects to prepare internal hooks at checkpoint */
+	acquireExclusiveVMAccess(currentThread);
+	vm->memoryManagerFunctions->j9mm_iterate_all_objects(vm, vm->portLibrary, 0, objectIteratorCallback, vm->checkpointState.hookRecords);
+
+	J9InternalHookRecord *hookRecord = (J9InternalHookRecord*)pool_startDo(hookRecords, &walkState);
+	while (NULL != hookRecord) {
+		if (!hookRecord->isRestore) {
+			if (FALSE == hookRecord->hookFunc(vm, hookRecord)) {
+				result = FALSE;
+				break;
+			}
+		}
+		hookRecord = (J9InternalHookRecord*)pool_nextDo(&walkState);
+	}
+	Trc_VM_criu_runCheckpointHooks_Exit(currentThread);
+
+	return result;
+}
+
+BOOLEAN
+runInternalJVMRestoreHooks(J9JavaVM *vm)
+{
+	J9VMThread *currentThread = vm->mainThread;
+	J9Pool *hookRecords = vm->checkpointState.hookRecords;
+	pool_state walkState = {0};
+	BOOLEAN result = TRUE;
+
+	Trc_VM_criu_runRestoreHooks_Entry(currentThread);
+	J9InternalHookRecord *hookRecord = (J9InternalHookRecord*)pool_startDo(hookRecords, &walkState);
+	while (NULL != hookRecord) {
+		if (hookRecord->isRestore) {
+			if (FALSE == hookRecord->hookFunc(vm, hookRecord)) {
+				result = FALSE;
+				break;
+			}
+		}
+		hookRecord = (J9InternalHookRecord*)pool_nextDo(&walkState);
+	}
+
+	/* Cleanup at restore */
+	releaseExclusiveVMAccess(currentThread);
```

Comment:
we dont have to release exclusive here, i think its more clear to handle exclusive access in teh jcl native

Reply:
Moved it into `Java_org_eclipse_openj9_criu_CRIUSupport_checkpointJVMImpl`.